mono/tests/test-runner.cs

```diff
@@ -154,182 +143,145 @@ class TestInfo {
+					info.RedirectStandardError = true;
+					Process p = new Process ();
+					p.StartInfo = info;
+
+					ProcessData data = new ProcessData ();
+					data.test = test;
+
+					string log_prefix = "";
+					if (opt_set != null)
+						log_prefix = "." + opt_set.Replace ("-", "no").Replace (",", "_");
+
+					data.stdoutFile = test + log_prefix + ".stdout";
+					data.stdout = new StreamWriter (new FileStream (data.stdoutFile, FileMode.Create));
+
+					data.stderrFile = test + log_prefix + ".stderr";
+					data.stderr = new StreamWriter (new FileStream (data.stderrFile, FileMode.Create));
+
+					p.OutputDataReceived += delegate (object sender, DataReceivedEventArgs e) {
+						if (e.Data != null) {
+							data.stdout.WriteLine (e.Data);
+						} else {
+							data.stdout.Flush ();
+							data.stdout.Close ();
+						}
+					};
+
+					p.ErrorDataReceived += delegate (object sender, DataReceivedEventArgs e) {
+						if (e.Data != null) {
+							data.stderr.WriteLine (e.Data);
+						} else {
+							data.stderr.Flush ();
+							data.stderr.Close ();
+						}
+					};
+
+					p.Start ();
+
+					p.BeginOutputReadLine ();
+					p.BeginErrorReadLine ();
+
+					if (!p.WaitForExit (timeout * 1000)) {
+						lock (monitor) {
+							timedout.Add (data);
+							failed.Add (data);
+						}
 
-			ThreadPool.QueueUserWorkItem (o => {
-				Process process = (Process) o;
+						if (concurrency == 1)
+							Console.WriteLine ("timed out.");
+						else
+							Console.Write (".");
 
-				process.WaitForExit ();
+						p.Kill ();
+					} else if (p.ExitCode != expectedExitCode) {
+						lock (monitor) {
+							failed.Add (data);
+						}
 
-				lock (monitor) {
-					if (process.ExitCode == expectedExitCode) {
 						if (concurrency == 1)
-							Console.WriteLine ("passed.");
+							Console.WriteLine ("failed.");
 						else
 							Console.Write (".");
-						passed.Add(process_data [process]);
-						npassed ++;
 					} else {
+						lock (monitor) {
+							passed.Add (data);
+						}
+
 						if (concurrency == 1)
-							Console.WriteLine ("failed.");
+							Console.WriteLine ("passed.");
 						else
-							Console.Write ("F");
-						failed.Add (process_data [process]);
-						nfailed ++;
+							Console.Write (".");
 					}
-					processes.Remove (process);
-					terminated.Add (process);
-					Monitor.Pulse (monitor);
 				}
-			}, p);
+			});
+
+			thread.Start ();
+
+			threads.Add (thread);
 		}
 
-		bool timed_out = false;
+		for (int j = 0; j < threads.Count; ++j)
+			threads [j].Join ();
 
-		/* Wait for all processes to terminate */
-		while (true) {
-			lock (monitor) {
-				int nprocesses = processes.Count;
+		int npassed = passed.Count;
+		int nfailed = failed.Count;
+		int ntimedout = timedout.Count;
 
-				if (nprocesses == 0)
-					break;
+		List<ProcessData> failedortimedout = new List<ProcessData> (nfailed + ntimedout);
+		failedortimedout.AddRange (failed);
+		failedortimedout.AddRange (timedout);
```

Comment:
Instead of combining this into a single List (which is only used for the nunit xml), I think it's better to keep them separate. The nunit xml should also indicate whether a test failed or timed out.